target/cheri-common/cheri_tagmem.c

```diff
@@ -519,117 +511,149 @@ void cheri_tag_set(CPUArchState *env, target_ulong vaddr, int reg, hwaddr* ret_p
+        *prot = 0;
+        if (tagmem_flags & TLBENTRYCAP_FLAG_CLEAR)
+            *prot |= PAGE_LC_CLEAR;
+        if (tagmem_flags & TLBENTRYCAP_FLAG_TRAP)
+            *prot |= PAGE_LC_TRAP;
+    }
+
+    // Squash happens in the caller, so read the tagblk even if
+    // TLBENTRYCAP_FLAG_CLEAR
+    bool result =
+        (tagmem == ALL_ZERO_TAGBLK)
+            ? 0
+            : tagblock_get_tag_tagmem(tagmem, page_vaddr_to_tag_offset(vaddr));
+
     // XXX: Not atomic w.r.t. writes to tag memory
-    qemu_maybe_log_instr_extra(env, "    Cap Tag Read [" TARGET_FMT_lx "/"
-        RAM_ADDR_FMT "] -> %d\n", vaddr, (ram_addr_t)(tag * CHERI_CAP_SIZE), result);
+    qemu_maybe_log_instr_extra(
+        env, "    Cap Tag Read [" TARGET_FMT_lx "/" RAM_ADDR_FMT "] -> %d\n",
+        vaddr, qemu_ram_addr_from_host(host_addr), result);
     return result;
 }
 
-/*
- * QEMU currently tells the kernel that there are no caches installed
- * (xref target/mips/translate_init.inc.c MIPS_CONFIG1 definition)
- * so we're kind of free to make up a line size here.  For simplicity,
- * we pretend that our cache lines always contain 8 capabilities.
- */
-#define CAP_TAG_GET_MANY_SHFT    3
 int cheri_tag_get_many(CPUArchState *env, target_ulong vaddr, int reg,
         hwaddr *ret_paddr, uintptr_t pc)
 {
-    uint64_t tag;
-    int prot;
-    CheriTagBlock *tagblk =
-        cheri_tag_get_block(env, vaddr, MMU_DATA_CAP_LOAD, reg,
-                            CAP_TAG_GET_MANY_SHFT, pc, ret_paddr, &tag, &prot);
+
+    const int mmu_idx = cpu_mmu_index(env, false);
+    probe_read(env, vaddr, 1, mmu_idx, pc);
+    handle_paddr_return(read);
+
+    uintptr_t tagmem_flags;
+    void *tagmem =
+        get_tagmem_from_iotlb_entry(env, vaddr, mmu_idx, false, &tagmem_flags);
+
+    int result =
+        ((tagmem == ALL_ZERO_TAGBLK) || (tagmem_flags & TLBENTRYCAP_FLAG_CLEAR))
+            ? 0
+            : tagblock_get_tag_many_tagmem(tagmem,
+                                           page_vaddr_to_tag_offset(vaddr));
 
     /*
-     * XXX Right now, the sole consumer of this function is CLoadTags, and
-     * we let it "see around" the TLB capability load inhibit.  That should
-     * perhaps change?
+     * The Mips CLoadTags, can "see around" the TLB capability load inhibit.
+     * That should perhaps change?
      */
-    (void) prot;
-
-    if (tagblk == NULL)
-        return 0;
-    else {
-#define TAG_BYTE_TO_BIT(ix)                                                    \
-    (tagblock_get_tag(tagblk, CAP_TAGBLK_IDX(tag + ix)) ? (1 << ix) : 0)
-        return TAG_BYTE_TO_BIT(0)
-             | TAG_BYTE_TO_BIT(1)
-             | TAG_BYTE_TO_BIT(2)
-             | TAG_BYTE_TO_BIT(3)
-             | TAG_BYTE_TO_BIT(4)
-             | TAG_BYTE_TO_BIT(5)
-             | TAG_BYTE_TO_BIT(6)
-             | TAG_BYTE_TO_BIT(7);
-#undef TAG_BYTE_TO_BIT
+#ifndef TARGET_MIPS
+    // FIXME: Should there be a load trap if this instruction is used and no
+    // tags returned?
```

Comment:
Hm.  Data dependence again.  I think it's fine to make this one unconditional if that simplifies actual processor pipelines (@PeterRugg may have an informed opinion here?).

But whatever it is we do we probably shouldn't have MIPS behaving differently.

Reply:
Definitely tricky for our pipelines, but potentially not a deal breaker if it turns out to make a big difference: more investigation needed I think

Reply:
In sail-cheri-riscv we currently trap if the PTE lacks CR, and with the revocation 3.0 PR it also does the same for generation mismatches (the trap isn't data-dependent, just `ptw_info.ptw_lc != PTW_LC_OK`). I see it as similar to the fact we make a lack of Permit_Load_Cap an error, though one could argue it's not the same as page table bits are rather orthogonal and, unlike Permit_Load_Cap where (a) there could be tags set that you could read with a different cap (b) you silently clear the tags, here instead all pointers to the memory will behave the same, and the PTE bits could change at any time. Hm. I suspect we should fix the Sail and make this behave more like LC, but for now this is fine and matches the Sail.

Reply:
Do we think its OK for me to get rid of the ifdef and make it 
`    if (result && (tagmem_flags & TLBENTRYCAP_FLAG_TRAP)) {
        raise_load_tag_exception(env, vaddr, reg, pc);
    }`
This would semantically match what multiple capability sized loads would do.